etest/build.py

```diff
@@ -1,32 +1,114 @@
-# The following could and should be excised to an architecture module that
-# handles the nuances of architecture differences in Gentoo and Docker.
 @click.option(
     "--architecture",
     "--arch",
-    type=click.Choice(["amd64", "x86", "arm64", "armv5", "armv7", "ppc64"], case_sensitive=False),
+    type=click.Choice(["amd64", "x86", "arm64", "armv5", "armv6", "armv7", "ppc64"], case_sensitive=False),
     default="amd64",
     help="Architecture for the built image.",
 )
-@click.option("--environment", "--env", multiple=True, help="Pass an environment variable through to Docker.")
-@click.option("--name", type=Path, default=os.path.basename(sys.argv[0]), help="Name of the directory?")
-def main(hardened: bool, multilib: bool, systemd: bool, architecture: str, environment: List[str], name: Path) -> None:
-    """Build the etest image."""
-    click.echo(f"hardened: {hardened}")
-    click.echo(f"multilib: {multilib}")
-    click.echo(f"systemd: {systemd}")
-    click.echo(f"architecture: {architecture}")
-    click.echo(f"environment: {environment}")
-    click.echo(f"name: {name}")
+@click.option(
+    "--libc",
+    type=click.Choice(["glibc", "musl", "uclibc"], case_sensitive=False),
+    default="glibc",
+    help="libc for the built image.",
+)
+@click.option("--path", type=Path, default=os.getcwd(), help="Path to the directory containing the Dockerfile.")
+def main(
+    quiet: bool,
+    verbose: bool,
+    hardened: bool,
+    multilib: bool,
+    systemd: bool,
+    architecture: str,
+    libc: str,
+    path: Path,
+) -> None:
+    """Build the etest image/s."""
+    if not (path / "Dockerfile").is_file():
+        raise FileNotFoundError(f"No Dockerfile found on {path}")
+
+    profile = Profile(quiet, architecture, libc, hardened, multilib, systemd)
+    profile.build()
+
+    if verbose:
+        click.echo(f"Current profile: {profile.profile}")
+
+    with qemu.qemu(profile.arch):
+        build(quiet, verbose, profile, path)
+
+
+def build(quiet: bool, verbose: bool, profile: Profile, path: Path) -> None:
+    """Build the image."""
+    try:
+        docker.image.build(
+            path=str(path),
+            buildargs={"PROFILE": profile.docker_profile},
+            tag=f"etest/stage1:{profile.profile}",
+        )
+
+        stage2 = docker.container.run(
+            image=f"etest/stage1:{profile.profile}",
+            command=commands[profile.libc],
+            privileged=True,
+            name=f"stage2-{profile.profile}",
+        )["Container"]
+
+        docker.container.commit(container=stage2, repository="alunduil/etest", tag=profile.profile)
+    except Exception as e:
+        click.echo("Exception triggered. Aborting.")
+
+        raise (e)
+    finally:
+        try:
+            # There's no problem if this errors, so leave it be. However, this specific order ensures cleanup always happens.
+            docker.image.remove(image=f"etest/stage1:{profile.profile}")
```

Comment:
I dont think I quite get the implementation you are suggesting, alltough on-the-air it seems reasonable, I cant imagine how to implement it. Mind sending a simple mockup via matrix?

Reply:
Sure, let's try this first though.  Let me know if you still want matrix for this if it's still unclear.

```python
try:
    stage1 = ...
finally:
    # cleanup stage1

try:
    stage2 = ...
finally:
    # cleanup stage2

Reply:
I dont think this will work, since stage2 uses stage1 as an image, so it cant build after we clean it up. Do you have any other ideas?

Reply:
Rather than a finally make cleanup an exception handler?  I'd just like to see cleanup happening when errors occur that doesn't need any strange conditional checks.  It'd probably shake out nicely if you factors out stage1 and stage2 into functions with corresponding cleanup functions.  Then the story would be super clear.

Reply:
That would be a good idea, but youd have to have `try` blocks inside the exception handler itself. I cant see how to factor them into separate functions too without having the same problem I talked about earlier

Reply:
Alright, let's go with working for now and I'll see if I can play with it after it lands and see what we end up with.  Can you add some basic test coverage for this function to get us started as well?